---
title: 'Projects'
date: 2022-09-01T14:35:21+01:00
draft: false
---

## Nextflow pipelines

### OrthoQuant

![OrthoQuant](/images/OrthoQuant_logo.png)

[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.10.3-23aa62.svg?labelColor=000000)](https://www.nextflow.io/)

<!-- [![run with conda](http://img.shields.io/badge/run%20with-conda-3EB049?labelColor=000000&logo=anaconda)](https://docs.conda.io/en/latest/) -->

[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed.svg?labelColor=000000&logo=docker)](https://www.docker.com/)
[![run with singularity](https://img.shields.io/badge/run%20with-singularity-1d355c.svg?labelColor=000000)](https://sylabs.io/docs/)

**[OrthoQuant](https://github.com/adamd3/OrthoQuant)** is a Nextflow pipeline for performing strain-specific bacterial RNA-Seq analysis without a reference genome. It is written in Python, Groovy, and R.

### BactSeq

[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.10.3-23aa62.svg?labelColor=000000)](https://www.nextflow.io/)

<!-- [![run with conda](http://img.shields.io/badge/run%20with-conda-3EB049?labelColor=000000&logo=anaconda)](https://docs.conda.io/en/latest/) -->

[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed.svg?labelColor=000000&logo=docker)](https://www.docker.com/)
[![run with singularity](https://img.shields.io/badge/run%20with-singularity-1d355c.svg?labelColor=000000)](https://sylabs.io/docs/)

**[BactSeq](https://github.com/adamd3/BactSeq)** is a Nextflow pipeline for analysing bacterial RNA-Seq data using a reference genome and an arbitrary number of conditions. It is written in Python, Groovy, and R.

## R packages

### codondiffR

![codondiffR](/images/codondiffR_small.png)

**[codondiffR](https://github.com/adamd3/codondiffR)** is an R package for the calculation, visualisation, and comparative analysis of codon usage metrics in user-supplied protein-coding nucleotide sequences. The package is designed primarily for the analysis of viral open reading frames (ORFs), with the intention of identifying potential host taxa based on overall codon usage similarity, but it can be used for the comparison of any protein-coding input sequence(s).

You can find an Introductory Vignette describing the main features of the package [here](https://adamd3.github.io/posts/codondiffr/codondiffr/).

## Web apps

### Cystic Fibrosis gene expression browser

![CF-browser](/images/cf-browser-screenshot.png)

The [Cystic Fibrosis gene expression browser](https://blueprint.haem.cam.ac.uk/cf-express/) provides access to multiple gene expression and epigenetic data sets generated as part of the Varsity Project at the University of Cambridge. It is hosted by the [Department of Haematology](https://www.haem.cam.ac.uk/).

Gene and transcript expression can be visualised and accessed directly via the app, and genome-wide H3K27Ac acetylation data are provided in the [UCSC browser](https://genome.ucsc.edu/).
